internal/compiler/wasm/optimizations.go

```diff
@@ -101,3 +105,150 @@ func withControlInstr(is []instruction.Instruction) bool {
 	}
 	return false
 }
+
+func unquote(s string) (string, error) {
+	return strconv.Unquote("\"" + strings.ReplaceAll(s, `\`, `\x`) + "\"")
+}
+
+func (c *Compiler) removeUnusedCode() error {
+	cgCSV, err := opa.CallGraphCSV()
+	if err != nil {
+		return fmt.Errorf("csv unpack: %w", err)
+	}
+	r := csv.NewReader(bytes.NewReader(cgCSV))
+	r.LazyQuotes = true
+	cg, err := r.ReadAll()
+	if err != nil {
+		return fmt.Errorf("csv read: %w", err)
+	}
+
+	cgIdx := map[uint32][]uint32{}
+	for i := range cg {
+		callerName, err := unquote(cg[i][0])
+		if err != nil {
+			return fmt.Errorf("unquote caller name %s: %w", cg[i][0], err)
+		}
+		calleeName, err := unquote(cg[i][1])
+		if err != nil {
+			return fmt.Errorf("unquote callee name %s: %w", cg[i][1], err)
+		}
+		caller, ok := c.funcs[callerName]
+		if !ok {
+			return fmt.Errorf("caller not found: %s (%s)", cg[i][0], callerName)
+		}
+		callee, ok := c.funcs[calleeName]
+		if !ok {
+			return fmt.Errorf("callee not found: %s (%s)", cg[i][1], calleeName)
+		}
+		cgIdx[caller] = append(cgIdx[caller], callee)
+	}
+
+	keepFuncs := map[uint32]struct{}{}
+	keep := func(i uint32) {
+		keepFuncs[i] = struct{}{}
+	}
+
+	// we'll keep
+	// - what's referenced in a table (these could be called indirectly)
+	// - what's exported or imported
+	// - what's been compiled by us
+	// - anything transitively called from those
+
+	for _, imp := range c.module.Import.Imports {
+		if _, ok := imp.Descriptor.(module.FunctionImport); ok {
+			keep(c.funcs[imp.Name])
+		}
+	}
+
+	for _, exp := range c.module.Export.Exports {
+		if exp.Descriptor.Type == module.FunctionExportType {
+			keep(c.funcs[exp.Name])
+		}
+	}
+
+	for _, f := range c.funcsCode {
+		fidx := c.funcs[f.name]
+		keep(fidx) // add to starter set
+
+		cgIdx[fidx] = findCallees(f.code.Func.Expr.Instrs)
+	}
+
+	// anything referenced in a table
+	for _, seg := range c.module.Element.Segments {
+		for _, idx := range seg.Indices {
+			keep(idx)
+		}
+	}
+
+	var keeps []uint32
+	seen := map[uint32]bool{}
+	for idx := range keepFuncs {
+		reach(cgIdx, &keeps, seen, idx)
+	}
+	for _, k := range keeps {
+		keep(k)
+	}
```

Comment:
This looks fine to me.

 I think it might be possible to simply have one set (`keepFuncs`) and define `reach` as `func(cg map[uint32][]uint32, keepFuncs map[uint32]struct{}, node uint32)`. The `reach` function would traverse from `node` to all reachable nodes in `cg` adding them to `keepFuncs`. If a node was already contained in `keepFuncs` it would not traverse further. Then we can just call `reach` in each of the places above where `keep` is currently being called. Then we don't need a few of these extra variables?

Reply:
Yep that should work. Let's do that.